src/main/java/mate/academy/AsyncRequestProcessor.java

```diff
@@ -1,16 +1,49 @@
 package mate.academy;
 
+import java.util.Map;
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
 import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
 
 public class AsyncRequestProcessor {
+    private final Map<String, UserData> cache = new ConcurrentHashMap<>();
     private final Executor executor;
 
     public AsyncRequestProcessor(Executor executor) {
         this.executor = executor;
     }
 
     public CompletableFuture<UserData> processRequest(String userId) {
-        return null;
+        if (userId == null || userId.isEmpty()) {
+            return CompletableFuture.failedFuture(
+                    new IllegalArgumentException("userId must not be null or empty"));
+        }
+
+        if (cache.containsKey(userId)) {
+            return CompletableFuture.completedFuture(cache.get(userId));
+        }
+
+        return CompletableFuture.supplyAsync(() -> {
+            sleep(500);
+            return cache.computeIfAbsent(userId, key -> new UserData(key, "Details for " + key));
+        }, executor);
```

Comment:
the `containsKey()` check is redundant here because you use `computeIfAbsent()` down below which does it for you